docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+  - [HttpOperationException]: Create a new exception called HttpOperationException, which includes a StatusCode property, and implement the necessary logic to map the exception from HttpStatusCode, HttpRequestException, or Azure.RequestFailedException. Update existing SK code that interacts with the HTTP stack to throw HttpOperationException in case of a failed HTTP request and assign the original exception as its inner exception.
+
+- **Other**:
+  - Remove any code that wraps unhandled exceptions into AIException or any other SK exception solely for the purpose of wrapping. In most cases, this code does not provide useful information to help SK client code handle the exception, apart from a generic and uninformative "Something went wrong" message. Regardless of whether the unhandled exception is wrapped or not, SK client code cannot handle or recover from it without prior knowledge of the exception and proper exception handling logic.
+
+## Questions and consideration
+
+- Should the exception hierarchy follow the SK components hierarchy? For example, should there be a PlanningException for planners and a MemoryStorageException for memory storage connectors? What are the benefits for SK client code? Can the SK client code operate on the exception?  
+  In order to handle an exception successfully, the client code needs to know as many details as possible about the failed operation, such as exception type, status code, and some other operation related properties. This suggests that specific exception types with strongly typed properties should be preferred over generic/common exception types. If that's the case, what is the purpose of having base generic exceptions?  
+  On the other hand, having base exception types per component or set of related operations can be convenient from a code reusability perspective because derived types can inherit common component exception properties relevant to that component.
+- Should all custom SK exceptions be derived from the SKException type?  Currently, SKException derives from the Exception type and does not add any additional information. What value can SK client code get from catching these exceptions compared to catching them as Exception types? Should SKException be used in all cases when SK needs to indicate that something went wrong without providing details through strongly typed properties, and neither custom nor standard exception types exist that could be used for that purpose?
+- Should we keep the IsCriticalException functionality that SK currently has to avoid handling several system/critical exceptions such as OutOfMemoryException and StackOverflowException? Some of them cannot be intercepted.
```

Comment:
StackOverflowException isn't actually thrown; if you stack overflow, you crash, and the stack isn't unwound, e.g. your catch / finally blocks don't execute.  There's no benefit in trying to special-case it.

And OutOfMemoryExceptions don't necessarily mean you can't run recovery code.  The failure might be due, for example, to code trying to allocate an exorbitant amount of memory even though a smaller allocation might have succeeded.

Especially with moving to a model where exceptions aren't caught and stored, I don't think it's valuable to try to special-case certain exception types.

Reply:
Yep, removed this question and added a step to remove those two exceptions.